### PR TITLE
Fix 4689 Exclude ``ThreadPoolExecutor`` and ``ProcessPoolExecutor`` from ``consider-using-with``

### LINKED ISSUE #4689: `consider-using-with` on ThreadPoolExecutor?
## Question

It's pretty common to construct a `concurrent.futures.ThreadPoolExecutor` and keep it around as a singleton or a global variable or the like. Until recently this passed pylint just fine, but recently it's started giving me the error:

    R1732: Consider using 'with' for resource-allocating operations (consider-using-with)

Given that `ThreadPoolExecutor` typically is intended to sit around as a task queue that persists in the background, it seems strange for that one to be called out as a suggested use of `with`. Is there a particular pattern I should be following, or is this just a case where I should be suppressing the warning on a case-by-case basis?

### PR DESCRIPTION
**Pull request by DudeNr33**
## Steps

- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Description
This PR addresses two issues:
* Completely exclude ``ThreadPoolExecutor`` and ``ProcessPoolExecutor`` from ``consider-using-with``, as #4689 points out that legitimate use cases exist where they are used without a ``with`` block. 
* Check if a context manager that was assigned to a variable is used later on in a ``with`` block. Therefore, the following (although unusual, there are probably better examples for this) will no longer emit a ``consider-using-with`` message:
```python
file_handle = open("foo")  # formerly triggered ``consider-using-with``, not anymore now
with file_handle:
    content = file_handle.read()
```

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
| ✓   | :hammer: Refactoring   |

## Related Issue

Closes #4689

### CODE REVIEW

File: pylint/checkers/refactoring/refactoring_checker.py [hunk 1]
```diff
@@ -1317,9 +1384,58 @@ def visit_assign(self, node):
             )
         self.add_message(message, node=node, args=(suggestion,))
 
-    visit_return = visit_assign
+    def _append_context_managers_to_stack(self, node: astroid.Assign):
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/refactoring/refactoring_checker.py:1387]:
```suggestion
    def _append_context_managers_to_stack(self, node: astroid.Assign) -> None:
```

**Review by Pierre-Sassoulas — APPROVED** [review]
Thank you for taking the time to fix the false positive for this checker and participating in the issue related to it, much appreciated ! I made some suggestion feel free to ignore if it's not relevant.

**Comment by DudeNr33** [comment]
Thanks for the review, I will implement the suggested changes (but only tomorrow).
Especially the ``__iter__`` and a ``clear`` (or ``clear_all``) method are nice improvements.

File: pylint/checkers/refactoring/refactoring_checker.py [hunk 2]
```diff
@@ -1317,9 +1384,58 @@ def visit_assign(self, node):
...
+                return
+            values = value.elts
+        else:
+            assignees = [node.targets[0]]
+            values = [node.value]
+        if assignees is Uninferable or values is Uninferable:
+            return
+        for assignee, value in zip(assignees, values):
+            if isinstance(value, astroid.Call):
+                inferred = utils.safe_infer(value.func)
```
[the remark this answers is not in the document]
Comment (DudeNr33) [reply pylint/checkers/refactoring/refactoring_checker.py]: Resolved with 42132f5

File: pylint/checkers/refactoring/refactoring_checker.py [hunk 3]
```diff
@@ -593,7 +619,18 @@ def visit_excepthandler(self, node):
...
     @utils.check_messages("redefined-argument-from-local")
     def visit_with(self, node):
-        for _, names in node.items:
+        for var, names in node.items:
+            if isinstance(var, astroid.Name):
+                for stack in (
+                    self._consider_using_with_stack.function_scope,
+                    self._consider_using_with_stack.class_scope,
+                    self._consider_using_with_stack.module_scope,
+                ):
```
[the remark this answers is not in the document]
Comment (DudeNr33) [reply pylint/checkers/refactoring/refactoring_checker.py]: Resolved with 42132f5

File: pylint/checkers/refactoring/refactoring_checker.py [hunk 4]
```diff
@@ -425,6 +443,9 @@ def _init(self):
         self._nested_blocks_msg = None
         self._reported_swap_nodes = set()
         self._can_simplify_bool_op = False
+        self._consider_using_with_stack.module_scope.clear()
+        self._consider_using_with_stack.class_scope.clear()
+        self._consider_using_with_stack.function_scope.clear()
```
[the remark this answers is not in the document]
Comment (DudeNr33) [reply pylint/checkers/refactoring/refactoring_checker.py]: Resolved with 42132f5

**Comment by Pierre-Sassoulas** [comment]
Thanks for the fix ! This was the last MR for pylint 2.9.4 I'm going to release it now :)

**Pull request merged**